### PR TITLE
API-599: replace integration tests by end-to-end tests

### PR DESCRIPTION
**Pull request by Doodoune**
Replace all integration tests by end-to-end tests using "donatj/mock-webserver".

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+language: php
+
+php:
+    - 7.1
```
Comment (ahocquard) [inline .travis.yml]: 5.6
7.0

File: tests/Api/ApiTestCase.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                "refresh_token" : "this-is-a-refresh-token",
+                "access_token" : "this-is-an-access-token"
+            }
+JSON;
+    }
+
+    /**
+     * @return StreamFactory
+     */
+    protected function getStreamFactory()
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: where is it used?

File: tests/Api/CreateProductMediaFileTest.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        $mediaFile = realpath(__DIR__ . '/../fixtures/akeneo.png');
+
+        $response = $api->create($mediaFile, [
+            'identifier' => 'medium_boot',
+            'attribute'  => 'side_view',
+            'scope'      => null,
+            'locale'     => null,
+        ]);
+
+        $this->assertSame('f/b/0/6/fb068ccc9e3c5609d73c28d852812ba5faeeab28_akeneo.png', $response);
```
Comment (ahocquard) [inline tests/Api/CreateProductMediaFileTest.php]: Assert::assertSame

Comment (ahocquard) [reply tests/Api/CreateProductMediaFileTest.php]: Check in the history of the response stack (take a look at the library documentation) what if the body of the request what what we expects. 
For now, you are just checking the URI call.

File: tests/Api/CreateProductTest.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            '/'. sprintf(ProductApi::PRODUCTS_URI, 'black_sneakers'),
+            new ResponseStack(
+                new Response('', [], 201)
+            )
+        );
+
+        $api = $this->createClient()->getProductApi();
+        $response = $api->create('new_shoes', $this->newProduct());
+
+        $this->assertSame(201, $response);
```
Comment (ahocquard) [inline tests/Api/CreateProductTest.php]: same (almost evrywhere): check the body of the request. You can do a function in `ApiTestCase`.

File: tests/Api/ListPerPageProductTest.php [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+            new ResponseStack(
+                new Response($this->getFirstPage(), [], 200),
+                new Response($this->getSecondPage(), [], 200)
+            )
+        );
+
+        $api = $this->createClient()->getProductApi();
+        $firstPage = $api->listPerPage(10, true, []);
+
+        $this->assertInstanceOf(PageInterface::class, $firstPage);
```
Comment (ahocquard) [inline tests/Api/ListPerPageProductTest.php]: Assert::

File: tests/Api/ApiTestCase.php (hunk 2 above)
Comment (Doodoune) [reply tests/Api/ApiTestCase.php]: In UpsertListProductTest sur le test test_upsert_list_from_stream

File: .travis.yml (hunk 1 above)
Comment (Doodoune) [reply .travis.yml]: Added

File: tests/Api/ListPerPageProductTest.php (hunk 5 above)
Comment (Doodoune) [reply tests/Api/ListPerPageProductTest.php]: Changed everywere

File: composer.json [hunk 6]
```diff
@@ -27,13 +27,16 @@
     "php-http/discovery": "^1.0",
     "php-http/message-factory": "^v1.0",
     "php-http/multipart-stream-builder": "^1.0",
-    "php-http/client-implementation": "^1.0"
+    "php-http/client-implementation": "^1.0",
+    "php-http/guzzle6-adapter": "^1.1"
```
Comment (LaurentPetard) [inline composer.json]: Both in "require" and "require-dev" ?

**Comment by Doodoune** [comment]
@LaurentPetard I remove both, we add it by composer require in the doc so I don't have to add it in composer.json

**Comment by ahocquard** [comment]
Almost GTM, except for the BC break

File: tests/Api/ListPerPageProductTest.php [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        Assert::assertInstanceOf(PageInterface::class, $firstPage);
+        Assert::assertEquals($firstPage->getCount(), 11);
+        Assert::assertNull($firstPage->getPreviousLink());
+        Assert::assertNull($firstPage->getPreviousPage());
+        Assert::assertFalse($firstPage->hasPreviousPage());
+        Assert::assertTrue($firstPage->hasNextPage());
+        Assert::assertSame($this->server->getServerRoot() . '/api/rest/v1/products?page=2&with_count=true&pagination_type=page&limit=10', $firstPage->getNextLink());
+        Assert::assertEquals(count($firstPage->getItems()), 10);
+
+        $secondPage = $firstPage->getNextPage();
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/Api/ListPerPageProductTest.php]: It's already present a few line before : Assert::assertSame($this->server->getLastRequest()->jsonSerialize()[RequestInfo::JSON_KEY_GET], ['limit' => '10', 'with_count' => 'true']);

File: tests/Api/DownloadProductMediaFileTest.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+use donatj\MockWebServer\Response;
+use donatj\MockWebServer\ResponseStack;
+use PHPUnit\Framework\Assert;
+use Psr\Http\Message\StreamInterface;
+
+class DownloadProductMediaFileTest extends ApiTestCase
+{
+    public function test_download_media_file()
+    {
+        $expectedMediaFile = realpath(__DIR__ . '/../fixtures/akeneo.png');
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/Api/DownloadProductMediaFileTest.php]: Done

File: tests/Api/CreateProductTest.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            '/'. ProductApi::PRODUCTS_URI,
+            new ResponseStack(
+                new Response('', [], 201)
+            )
+        );
+
+        $api = $this->createClient()->getProductApi();
+        $response = $api->create('new_shoes', $this->newProduct());
+
+        Assert::assertSame($this->server->getLastRequest()->jsonSerialize()[RequestInfo::JSON_KEY_INPUT], json_encode(array_merge($this->newProduct(), ["identifier" => "new_shoes"])));
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/Api/CreateProductTest.php]: Done

File: src/Search/Operator.php [hunk 10]
```diff
@@ -27,7 +27,7 @@
     const GREATER_OR_EQUALS_THAN_ON_ALL_LOCALES = 'GREATER OR EQUALS THAN ON ALL LOCALES';
     const LOWER_THAN_ON_ALL_LOCALES = 'LOWER THAN ON ALL LOCALES';
     const LOWER_OR_EQUALS_THAN_ON_ALL_LOCALES = 'LOWER OR EQUALS THAN ON ALL LOCALES';
-    const EMPTY = 'EMPTY';
+    const IS_EMPTY = 'EMPTY';
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Search/Operator.php]: Done

File: phpunit.xml.dist [hunk 11]
```diff
@@ -7,22 +7,11 @@
      convertErrorsToExceptions="true"
      convertNoticesToExceptions="true"
      convertWarningsToExceptions="true"
-     processIsolation="true"
+     processIsolation="false"
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply phpunit.xml.dist]: Yeah, I change that

File: tests/Api/ListPerPageProductTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahocquard) [reply tests/Api/ListPerPageProductTest.php]: yes, but nextPage does another new request. It's a cursor under the hood. So you have to check that this new call is done with good query parameters

File: src/Search/Operator.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahocquard) [reply src/Search/Operator.php]: at the end, we will have to change it because it is not supported by php 5.6.... it's a kind of bug

**Pull request merged**